### PR TITLE
Add `auth::{Authorization, AuthenticationScheme, BasicAuth, WwwAuthenticate, ProxyAuthorization, ProxyAuthenticate}`

### PR DESCRIPTION
**Pull request by yoshuawuyts**
Ref #99 

Adds HTTP basic auth and related structs for authentication.

BasicAuth is a specialization of `Authorization`; because schemes like oauth are rather elaborate, we've chosen to put them into separate structs.

Thanks!

## Screenshots

Screenshot_2020-10-01 http_types auth - Rust

### CODE REVIEW

**Comment by yoshuawuyts** [comment]
I've added support for `WWW-Authenticate`, but need to change it up slightly before merging. Right now we support at most 1 authentication scheme, but the spec says this should be up to n. Also since `Proxy-Authenticate` parses *exactly* the same way as `WWW-Authenticate` we can copy-paste the impl once it works. That should provide the majority of common authentication things.

**Comment by yoshuawuyts** [comment]
Okay; all implemented!

**Comment by Fishrock123** [comment]
`BasicAuth` seems to work fine, as I am using it here: https://github.com/squamishaccess/squamishaccess-signup-function-rs/blob/6c445916760a4db21da80c732fbe8e9c704c1bd1/src/main.rs#L168-L172

However I did run into a UX issue with typed headers and surf: https://github.com/http-rs/surf/issues/251

File: src/auth/basic_auth.rs [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+        let bytes = base64::decode(auth.credentials()).status(400)?;
+        let credentials = String::from_utf8(bytes).status(400)?;
+
+        let mut iter = credentials.splitn(2, ':');
+        let username = iter.next();
+        let password = iter.next();
+
+        let (username, password) = match (username, password) {
+            (Some(username), Some(password)) => (username.to_string(), password.to_string()),
+            (Some(_), None) => bail!(400, "Expected basic auth to a password"),
```
Comment (Fishrock123) [inline src/auth/basic_auth.rs:76]:
```suggestion
            (Some(_), None) => bail!(400, "Expected basic auth to contain a password"),
```

File: src/macros.rs [hunk 2]
```diff
@@ -84,3 +84,99 @@ macro_rules! format_err {
         $crate::private::new_adhoc(format!($fmt, $($arg)*))
     };
 }
+
+/// Return early with an error and a status code.
+#[doc(hidden)]
+#[macro_export]
+macro_rules! bail2 {
```
Comment (Fishrock123) [inline src/macros.rs:91]: Maybe we should consider calling these new versions `{}_status`?

**Review by Fishrock123 — APPROVED** [review]
For some reason I had the impression this covered more than basic auth. Looks good, although I am not familiar with `WWW-Authenticate`.

Basic auth certainly works when I make client requests to mailchimp.

One question I have is how to make Authorization and Proxy-Authorization work seamlessly.

File: src/macros.rs (hunk 2 above)
Comment (yoshuawuyts) [reply src/macros.rs:91]: Sure thing! Though they're only internal for now which means the naming doesn't matter too much -- for the next major we should probably make these the default as per https://github.com/http-rs/http-types/issues/227

**Pull request merged**